docs/testing.rst

```diff
@@ -0,0 +1,140 @@
+        with app.test_client() as client:
+            with client.session_transaction() as sess:
+                sess['google_oauth_token'] = {
+                    'access_token': 'this is totally fake',
+                    'id_token': 'this is not a real token',
+                    'token_type': 'Bearer',
+                    'expires_in': '3600',
+                    'expires_at': fake_time + 3600,
+                }
+            yield client
+    
+This will inject the necessary information in order to ensure that Flask-Dance,
+requests-oauthlib and oauthlib2 believe there is a valid session, and one that
+won't need to be refreshed for another hour (3600 seconds). If your tests run
+longer than that, you'll need to adjust that value.
+
+The `fake_time` is created so that you can always refer to the same point in
+time throughout tests, and in any other fixture you might create. Alternatively
+you can use something like pytest-freezegun and then call any of the `time` and
+`datetime` functions as you normally would:
+
+.. code-block:: python
+
+    @pytest.fixture
+    @pytest.mark.freeze_time('2018-05-04')
+    def loggedin_app(app):
+        """Creates a logged-in test client instance."""
+        with app.test_client() as client:
+            with client.session_transaction() as sess:
+                sess['domain'] = 'example.com'
+                sess['google_oauth_token'] = {
+                    'access_token': 'this is totally fake',
+                    'id_token': 'this is not a real token',
+                    'token_type': 'Bearer',
+                    'expires_in': '3600',
+                    'expires_at': time.time() + 3600,
+                }
+            yield client
+
+Now that we have a logged-in client you can call any of the routes that are
+and check their responses.
+
+.. code-block:: python
+
+    def test_not_logged_in(app):
+        """Test that we redirect to Google to login."""
+        res = app.test_client().get('/')
+        assert ('redirected automatically to target URL: <a href="/login/'
+                'google">/login/google</a>').lower() in res.data.lower()
+        assert res.status_code == 302
+
+
+    def test_logged_in_index(loggedin_app):
+        """Tests getting the index route.
+
+        This will render the normal template as we have a valid oauth2 session.
+        """
+        res = loggedin_app.get('/')
+        assert res.content_type == 'text/html; charset=utf-8'
+        assert res.status_code == 200
+        assert 'something only shown when logged in' in res.data.lower()
+
+
+Calling authenticated APIs
+--------------------------
+
+Though we've managed to create a working session a problem now arises if you
+try to actually call an API, by using `google.get('some url')` for example.
+Your token will fail to validate and the request will be denied.
+
+This can be handled by a Python library called responses, which lets us control
```

Comment:
`responses` is a very useful library, and it's what I used to write tests for Flask-Dance itself. However, _after_ I wrote those tests, I found out about another library called [`betamax`](https://github.com/betamaxpy/betamax), which is written by one of the maintainers of Requests. It's a really nice library for intercepting HTTP requests and replaying them in tests, but it's more complicated to set up.

You might want to look into `betamax`, and see if it fits your usecase. If it does, and you're inspired to rewrite this example using `betamax` instead of `responses`, let me know! If not, or if you're happy with the documentation referring to `responses`, then that's fine -- I'm happy to get this pull request merged either way. (Just wanted to tell you about this cool library!)

Reply:
I had the same experience, became aware of betamax after doing this. I'd lprefer to submit a PR updating or adding a betamax example later on, once I've actually had some time to use it.

Reply:
Totally reasonable! Let me know if I can help or answer questions about Betamax and Flask-Dance.